[CASE safe_decorator_no_params]
[disable_cache]
from returns.result import safe

@safe
def test() -> int:
    return 1

reveal_type(test)  # E: Revealed type is 'def () -> returns.result.Result[builtins.int*, builtins.Exception]'
[/CASE]


[CASE safe_composition_no_params]
[disable_cache]
from returns.result import safe

def test() -> int:
    return 1

reveal_type(safe(test))  # E: Revealed type is 'def () -> returns.result.Result[builtins.int*, builtins.Exception]'
[/CASE]


[CASE safe_decorator_with_args]
[disable_cache]
from returns.result import safe

@safe
def test(first: int, second: str = None, *, kw: bool = True) -> int:
    return 1

reveal_type(test)  # E: Revealed type is 'def (first: builtins.int, second: Union[builtins.str, None] =, *, kw: builtins.bool =) -> returns.result.Result[builtins.int*, builtins.Exception]'
[/CASE]


[CASE safe_composition_with_args]
[disable_cache]
from returns.result import safe

def test(first: int, second: str = None, *, kw: bool = True) -> int:
    return 1

reveal_type(safe(test))  # E: Revealed type is 'def (first: builtins.int, second: Union[builtins.str, None] =, *, kw: builtins.bool =) -> returns.result.Result[builtins.int*, builtins.Exception]'
[/CASE]


[CASE safe_decorator_with_args_kwargs]
[disable_cache]
from returns.result import safe

@safe
def test(*args, **kwargs) -> int:
    return 1

reveal_type(test)  # E: Revealed type is 'def (*args: Any, **kwargs: Any) -> returns.result.Result[builtins.int*, builtins.Exception]'
[/CASE]


[CASE safe_decorator_with_typed_args_kwargs]
[disable_cache]
from returns.result import safe

@safe
def test(*args: int, **kwargs: str) -> int:
    return 1

reveal_type(test)  # E: Revealed type is 'def (*args: builtins.int, **kwargs: builtins.str) -> returns.result.Result[builtins.int*, builtins.Exception]'
[/CASE]
